packages/workspace/src/affine/index.ts

```diff
@@ -0,0 +1,161 @@
+import type { DatasourceDocAdapter } from '@affine/y-provider';
+import type { Socket } from 'socket.io-client';
+import { Manager } from 'socket.io-client';
+import {
+  applyAwarenessUpdate,
+  type Awareness,
+  encodeAwarenessUpdate,
+} from 'y-protocols/awareness';
+import type { Doc } from 'yjs';
+
+import {
+  type AwarenessChanges,
+  base64ToUint8Array,
+  uint8ArrayToBase64,
+} from './sync-socket-io/utils';
+
+const ioManager = new Manager(runtimeConfig.serverUrlPrefix + '/', {
+  autoConnect: false,
+});
+
+export const createAffineDataSource = (
+  id: string,
+  rootDoc: Doc,
+  awareness: Awareness
+) => {
+  if (id !== rootDoc.guid) {
+    console.warn('important!! please use doc.guid as roomName');
+  }
+
+  const socket = ioManager.socket('/');
+  socket.connect();
+
+  return {
+    get socket() {
+      return socket;
+    },
+    queryDocState: async (guid, options) => {
+      const stateVector = options?.stateVector
+        ? await uint8ArrayToBase64(options.stateVector)
+        : undefined;
+
+      return new Promise((resolve, reject) => {
+        socket.emit(
+          'doc-load',
+          {
+            workspaceId: rootDoc.guid,
+            guid,
+            stateVector,
+          },
+          (update: Error | string | null) => {
+            if (update instanceof Error) {
+              reject(update);
+              return;
+            }
+
+            resolve(update ? base64ToUint8Array(update) : false);
+          }
+        );
+      });
+    },
+    sendDocUpdate: async (guid: string, update: Uint8Array) => {
+      socket.emit('client-update', {
+        workspaceId: rootDoc.guid,
+        guid,
+        update: await uint8ArrayToBase64(update),
+      });
+
+      return Promise.resolve();
+    },
+    onDocUpdate: callback => {
+      const onUpdate = async (message: {
+        workspaceId: string;
+        guid: string;
+        update: string;
+      }) => {
+        if (message.workspaceId === rootDoc.guid) {
+          callback(message.guid, base64ToUint8Array(message.update));
+        }
+      };
+      socket.on('server-update', onUpdate);
+      const destroyAwareness = setupAffineAwareness(socket, rootDoc, awareness);
+
+      return () => {
+        socket.off('server-update', onUpdate);
+        destroyAwareness();
+        // socket.disconnect()
```

Comment:
just in case, are we good to keep this?
 https://github.com/toeverything/AFFiNE/blob/fe32eccc03f5c3f8ca899afa309f3fe53b0c10b9/packages/workspace/src/affine/sync-socket-io/index.ts#L192

then use `destroyAwareness().then(() => socket.disconnect())`